### PR TITLE
Refactor: Move Eth and Evm types from bus-mapping to a new sub-crate eth-types

### PR DESCRIPTION
**Pull request by scroll-dev**
Step-1 of **"Disable geth-utils for non-test mode"** PR (https://github.com/scroll-tech/zkevm-circuits/pull/70)

Related discussion in issue https://github.com/appliedzkp/zkevm-circuits/issues/272

### Summary

1. Move `eth_types.rs` of bus-mapping to `lib.rs` of  new sub-crate eth-types.
2. Move all files of `evm` of bus-mapping to `evm_types` of eth-types except `opcodes`.
3. Update code and test of bus-mapping and zkevm-circuits.

### CODE REVIEW

**Review by ed255 — CHANGES_REQUESTED** [review]
Thanks for the contribution!  Overall looks really good.  Please take a look at my comments, they are small details.

**Comment by ed255** [comment]
Also it seems that there are conflicts with main (probably from a recently merged PR) that you'll need to address.

File: eth-types/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "eth-types"
+version = "0.1.0"
+edition = "2021"
+authors = ["CPerezz <c.perezbaro@gmail.com>"]
```
[the remark this answers is not in the document]
Comment (CPerezz) [reply eth-types/Cargo.toml:5]: Of course! I suggested that a long time ago!!!!

File: eth-types/src/evm_types.rs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        self.inc();
+        pre
+    }
+}
+
+/// Wrapper type over `usize` which represents the global counter The purpose of
+/// the /// `RWCounter` is to enforce that each Opcode/Instruction and Operation
+/// is unique and just executed once.
+#[derive(Clone, Copy, Eq, PartialEq, PartialOrd, Ord)]
+pub struct RWCounter(pub usize);
```
[the remark this answers is not in the document]
Comment (CPerezz) [reply eth-types/src/evm_types.rs:59]: I do completely agree!

**Review by CPerezz — COMMENTED** [review]
LGTM modulo @ed255 's comments

File: eth-types/src/error.rs [hunk 3]
```diff
@@ -1,6 +1,6 @@
 //! Error module for the bus-mapping crate
 
-use crate::eth_types::{Address, GethExecStep, Word};
+use crate::{Address, GethExecStep, Word};
 use core::fmt::{Display, Formatter, Result as FmtResult};
 use ethers_providers::ProviderError;
 use std::error::Error as StdError;
```
[the remark this answers is not in the document]
Comment (silathdiir) [reply eth-types/src/error.rs:6]: I will try to fix it in another PR. Thanks.

[the remark this answers is not in the document]
Comment (ed255) [reply eth-types/src/error.rs:6]: Opened an issue for that https://github.com/appliedzkp/zkevm-circuits/issues/294

**Comment by silathdiir** [comment]
Hi @ed255 and @CPerezz , it seems that the CI is broken on `main` branch. Could you help on that if you have time? Thanks.

**Comment by HAOYUatHZ** [comment]
> Hi @ed255 and @CPerezz , it seems that the CI is broken on `main` branch. Could you help on that when you have time? Thanks.

@silathdiir wait for https://github.com/appliedzkp/zkevm-circuits/pull/296 being merged then it should work

**Comment by HAOYUatHZ** [comment]
Or you could cherry-pick https://github.com/appliedzkp/zkevm-circuits/pull/296 for now

**Review by CPerezz — APPROVED** [review]
LGTM

**Review by ed255 — APPROVED** [review]
LGTM!

**Pull request merged**